qiskit/transpiler/preset_passmanagers/builtin_plugins.py

```diff
@@ -160,6 +165,49 @@ def pass_manager(self, pass_manager_config, optimization_level=None) -> PassMana
                 )
             )
             init.append(CommutativeCancellation())
+            # skip peephole optimization before routing if target basis gate set is discrete,
+            # i.e. only consists of Cliffords that an user might want to keep
+            # use rz, sx, x, cx as basis, rely on physical optimziation to fix everything later one
+            stdgates = get_standard_gate_name_mapping()
+
+            def _is_one_op_non_discrete(ops):
+                """Checks if one operation in `ops` is not discrete, i.e. is parameterizable
+                Args:
+                    ops (List(Operation)): list of operations to check
+                Returns
+                    True if at least one operation in `ops` is not discrete, False otherwise
+                """
+                for op in ops:
+                    if isinstance(op, str):
+                        op = stdgates.get(op, None)
+
+                    if (
+                        op is None
+                        or not isinstance(op, Instruction)
+                        or op.name in _discrete_skipped_ops
+                    ):
+                        continue
+
+                    if len(op.params) > 0:
+                        return True
+                return False
+
+            target = pass_manager_config.target
+            basis = pass_manager_config.basis_gates
+            # consolidate gates before routing if the user did not specify a discrete basis gate, i.e.
+            # * no target or basis gate set has been specified
+            # * target has been specified, and we have one non-discrete gate in the target's spec
+            # * basis gates have been specified, and we have one non-discrete gate in that set
+            do_consolidate_blocks_init = target is None and basis is None
+            do_consolidate_blocks_init |= target is not None and _is_one_op_non_discrete(
+                target.operations
+            )
+            do_consolidate_blocks_init |= basis is not None and _is_one_op_non_discrete(basis)
+
+            if do_consolidate_blocks_init:
+                init.append(Collect2qBlocks())
+                init.append(ConsolidateBlocks())
+                init.append(Split2QUnitaries())
```

Comment:
The logic here seems intended to avoid doing an unsafe code path if we know the basis to be unsafe (i.e. one of the basis gates is continuous in a way that we understand).

I think we might want to flip it a little bit, especially with this merging this close to the rc; rather than assuming we will do the pass, I think we should assume that we _won't_, and only run it if we can prove the basis is made entirely of standard gates we understand and potentially one that isn't.

I'm thinking of the case where a downstream user has a custom gate that they've added to the basis translator's equivalences and they want to transpile against.  We have downstream packages like that (e.g. the circuit-knitting toolbox and some of our researchers), but either way it's part of our interface.

Reply:
I don't think there really is anything different being introduced by this and I viewed the if statement as unnecessary, see our earlier discussion thread here: https://github.com/Qiskit/qiskit/pull/12727#discussion_r1672182189 . Basically we either run `ConsolidateBlocks` here or we run it after routing/basis translation. From the preserving custom gates or non-standard gate workflows perspective nothing should be different by doing this. That being said I was fine leaving this in as extra insurance and given our proximity to the 1.2.0rc1 tag restricting it even further makes sense. We can loosen these requirements post 1.2.0.

Reply:
Ah yeah, it's a fair point that we do it after routing again anyway.  Yeah, I guess my concern here applies equally to our pre-existing situation - basically, my concern is that we should collect up gate if we can't be sure we'll be able to synthesise ourselves back out of them.

I'm ok leaving it in as-is, given we already run `ConsolidateBlocks` at these levels.

Reply:
I made the checks a bit stricter, I think this makes sense for a new pass that will essentially run on all input circuits, we can relax in a follow-up.